File: LinkedList /swapping.java
//You are given the head of a singly linked list. Your task is to swap the first and last nodes of the linked list and return the head of the modified list.
class Node {
    int data;
    Node next;
    Node(int d) {
        data = d;
        next = null;
    }
}

public class SwapFirstLast {

    static Node swapFirstLast(Node head) {
        // Base case: empty list or single node
        if (head == null || head.next == null) return head;

        // If only 2 nodes, just swap them
        if (head.next.next == null) {
            Node second = head.next;
            second.next = head;
            head.next = null;
            return second;
        }

        Node prevLast = head;
        while (prevLast.next.next != null) {
            prevLast = prevLast.next;
        }

        Node last = prevLast.next;
        prevLast.next = head;   // Last node points to second node via head later
        last.next = head.next;  // Last node's next = second node
        head.next = null;       // First node becomes last
        head = last;            // Update head to last node

        return head;
    }

    static void printList(Node head) {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + (temp.next != null ? "->" : "\n"));
            temp = temp.next;
        }
    }

    public static void main(String[] args) {
        Node head = new Node(10);
        head.next = new Node(20);
        head.next.next = new Node(30);
        head.next.next.next = new Node(40);
        head.next.next.next.next = new Node(50);

        System.out.println("Original List:");
        printList(head);

        head = swapFirstLast(head);

        System.out.println("Modified List:");
        printList(head);
    }
}

File: LinkedList /swapping.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def swap_first_last(head):
    if not head or not head.next:
        return head  # Empty or single node
    
    # If only 2 nodes
    if not head.next.next:
        second = head.next
        second.next = head
        head.next = None
        return second

    prev_last = head
    while prev_last.next.next:
        prev_last = prev_last.next

    last = prev_last.next
    prev_last.next = head
    last.next = head.next
    head.next = None
    head = last

    return head

def print_list(head):
    temp = head
    while temp:
        print(temp.data, end="->" if temp.next else "\n")
        temp = temp.next

# Example usage
head = Node(10)
head.next = Node(20)
head.next.next = Node(30)
head.next.next.next = Node(40)
head.next.next.next.next = Node(50)

print("Original List:")
print_list(head)

head = swap_first_last(head)

print("Modified List:")
print_list(head)
